### PR TITLE
Feature: Obtaining Crop Image Selections 

### PR DESCRIPTION
**Pull request by axiomcura**
# Image cropping workflow

In this PR, this introduces the very first feature that captures positional coordinates that represent a specific crop location within a source image file.

## Image cropping

cell_image_crop

> Stock cell image that show green boarders of where the images are going to be crop. Each individual square represents a cropped selection. 

Images are cropped by using a python package known as `Pillow`.

The images are cropped in evenly sized pieces (users can input the size of the crop). However, these cropped images are not saved into memory.  What are saved are objects known as `ImageCropSelection` objects.

With the example above, each square will be a single `ImageCropSelection` object that contains:

- `crop_id` -> Unique number appointed to crop data
- `file_name` -> points path to the original image
- `img_size` -> size of cropped image (L x W)
- `crop_position` -> Tuple object that contains the `X`, `Y`, `L` and `W` positions

These objects will serve as placeholders indicating where the cropping should be established from the main image source.

## Saving cropped images

```json
"cell_image_1.tiff": [
    {
        "img_id": 1,
        "file_name": "/Users/erikserrano/Development/Projects/macrophage-striation-analysis/notebooks/data/cell_image_1.tiff",
        "img_size": [
            256,
            256
        ],
        "crop_position": [
            0,
            0,
            256,
            256
        ]
    },
```

> Example of a JSON file containing crop information

Since the program does not save the cropped images into the local computer, the selection will be written out in a JSON file.

The JSON file contains crop information to be transferable to other applications, for example, to a Graphical User Interface (GUI) allowing it to find the image source, and where to crop it.

## additional information
- usage on how to load tiff image, cropping and writing into JSON file is found in the notebook folder
- Code functions are found within the `src` folder

### CODE REVIEW

File: notebooks/00_extract_coordinate_images.ipynb
Comment (gwaybio) [inline notebooks/00_extract_coordinate_images.ipynb:1]: Please make sure your notebooks are sequentially executed starting with cell [1]. (It looks like this one is out of order)

File: src/common/_io.py
Comment (gwaybio) [inline src/common/_io.py]: This has the potential to get _huge_.

I think it is ok to merge this in for now as is, but it is worth thinking about in the future if loading all images is necessary. Can you think of a way to not need all images in memory at any point?

File: src/common/_io.py
Comment (gwaybio) [inline src/common/_io.py]: it looks to me that the default is "image_cords"? Also, should it be "image_coords"? (two o's)?

File: src/structs/selections.py
Comment (gwaybio) [inline src/structs/selections.py:12]: nice! love it

File: src/utils/image_utils.py
Comment (gwaybio) [inline src/utils/image_utils.py:16]: Maybe worth bringing up now, but probably not worth it holding up merging: Why do we need to walk over an image? Is it ok for two images to be overlapping? Why not randomly walk, which might prevent us from needing to hold everything in memory?

**Review by gwaybio — APPROVED** [review]
This is a very elegant PR @axiomcura - well done!

I have minor comments that don't necessarily need to prevent merging. The comments are mostly meant to provoke thought

File: env.yaml
Comment (d33bs) [inline env.yaml:17]: Noticing the inclusion of pre-commit, is there a related `.pre-commit-config.yaml` file which should be checked in?

File: notebooks/00_extract_coordinate_images.ipynb
Comment (d33bs) [inline notebooks/00_extract_coordinate_images.ipynb]: The second sentence here looks like it may need minor adjustment.

File: src/common/_io.py
Comment (d33bs) [inline src/common/_io.py]: Should this error be thrown anywhere below?

File: src/common/_io.py
Comment (d33bs) [inline src/common/_io.py]: Underscores in module names are generally discouraged within Python, with prefix underscores sometimes commonly representing other things (see [PEP 8](https://peps.python.org/pep-0008/#package-and-module-names)). Is there a more descriptive name which could be provided to this module?

File: src/common/_io.py
Comment (d33bs) [inline src/common/_io.py]: Should this line be included in the loop above somehow (`crop_entry` is a loop variable)?

File: src/common/paths.py
Comment (d33bs) [inline src/common/paths.py]: It looks like this sentence may need more content.

File: src/common/paths.py
Comment (d33bs) [inline src/common/paths.py:12]: Consider adding a class docstring to describe the purpose of the class.

File: src/common/paths.py
Comment (d33bs) [inline src/common/paths.py]: Seeing this error message repeated, would it make sense to change it to mention something about the `.git` folder instead of the project root folder?

File: src/structs/selections.py
Comment (d33bs) [inline src/structs/selections.py:7]: Consider adding a module-level docstring to describe the purpose of this file.

File: src/structs/selections.py
Comment (d33bs) [inline src/structs/selections.py:21]: Consider adding a method docstring to describe the purpose.

File: src/utils/image_utils.py
Comment (d33bs) [inline src/utils/image_utils.py]: Should `x` and `y` be used somewhere below? If they aren't needed, consider removing this line.

File: src/utils/image_utils.py
Comment (d33bs) [inline src/utils/image_utils.py]: Minor suggestion here.

```suggestion
        # checking if the crop size is exact, if not skip
```

File: src/utils/image_utils.py
Comment (d33bs) [inline src/utils/image_utils.py]: Should this be an import from `src.structs.selections` instead?

```suggestion
from src.structs.selections import ImageCropSelection
```

**Review by d33bs — COMMENTED** [review]
Nice work! It was great seeing the `dataclass` and `TypeGuard` implementations. I left a few comments and suggestions with this review, please don't hesitate to let me know if you have any questions or if I may clarify.

One additional thought: your PR description and the image included were helpful for understanding the work included here. Would it make sense to reuse any of it within the repo's `readme.md` or other areas for documentation?

File: src/common/_io.py
Comment (axiomcura) [reply src/common/_io.py]: Yes, I was thinking about this when I was reviewing it myself. I plan on changing this into a python generator object instead of returning a list. ( it is on the works!)

File: src/common/_io.py
Comment (axiomcura) [reply src/common/_io.py]: Yikes, thanks for pointing that out! I'll fix that

File: env.yaml
Comment (axiomcura) [reply env.yaml:17]: Hey @d33bs , thanks for pointing this out. I do have a small question about this. Since I use pre commit hooks:
 - [isort](https://github.com/PyCQA/isort)
 - [black](https://github.com/psf/black) 
 - [sourcery](https://github.com/sourcery-ai/sourcery)

Should I leave the `.pre-commit-config.yaml` in the remote repo? This is my first time using git pre-commit hooks :)

File: src/utils/image_utils.py
Comment (axiomcura) [reply src/utils/image_utils.py:16]: Hey @gwaybio, good question. I am still learning on how to use `pillow` 😅 ! 

Once I have a better grasp on how `pillow` works, I may be able to answer your questions through a different PR!  For now, I will create an issue in regard to your comments and get back to it when I can implement these features!

Comment (axiomcura) [reply src/utils/image_utils.py:16]: Issue posted!: #2

File: src/common/_io.py
Comment (axiomcura) [reply src/common/_io.py]: This has been renamed to `file_io`, thanks @d33bs for the resource!

File: src/common/_io.py
Comment (axiomcura) [reply src/common/_io.py]: This was a hacky way of ensuring that each key was a file name of the source image. However, I have noticed that this is prone to some bugs; therefore, I have reimplemented this code block by using `defaultdicts`. 

Here is the commit: https://github.com/WayScience/macrophage-striation-analysis/pull/1/commits/47333b1a36f981d7990c546add2e3d3f38b52515

File: src/utils/image_utils.py
Comment (axiomcura) [reply src/utils/image_utils.py]: `x` and `y` are not used! Thanks for pointing this out!

**Comment by axiomcura** [comment]
Hello @gwaybio and @d33bs !

Thank you for all your comments! I have addressed some of the changes that requested from both of you. There are still some open conversations that I would like to be discussed before I merge this PR, please take a look at them!

Thanks!

File: env.yaml
Comment (d33bs) [reply env.yaml:17]: I would recommend including it. Including that file helps make sure it stays with the repository and can enable collaborators to use the same pre-commit configuration (unifying code style or quality expectations). Also, if you're including `black` as part of your `pre-commit` checks, you might consider removing it from the conda environment.

File: src/utils/image_utils.py
Comment (gwaybio) [reply src/utils/image_utils.py:16]: perfect.

**Comment by gwaybio** [comment]
> There are still some open conversations that I would like to be discussed before I merge this PR, please take a look at them!

I don't think I have any open conversations - you have my 👍 to merge!

File: src/common/_io.py
Comment (d33bs) [reply src/common/_io.py]: Looks great, nice use of `defaultdict`!

**Review by d33bs — APPROVED** [review]

**Comment by d33bs** [comment]
Ditto the above, great work on revisions! The pre-commit comment is a recommendation but not an unshifting requirement.

**Comment by axiomcura** [comment]
> Ditto the above, great work on revisions! The pre-commit comment is a recommendation but not an unshifting requirement.

Whoops! I forgot. I have added them now! 

Merging, thank you :)

**Pull request merged**
